### ISSUE TITLE
Semi Supervised Kernel Learning

### ISSUE THREAD

**Issue by hubert98** [cornellius-gp/gpytorch#1616]
Hi all,

I am currently trying to understand how to do semi-supervised learning with gpytorch based on this paper: https://arxiv.org/pdf/1805.10407.pdf

I would like to setup a mini-batch approach similar to this example: https://docs.gpytorch.ai/en/latest/examples/06_PyTorch_NN_Integration_DKL/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.html

Unfortunately I have no idea how to setup the unlabelled data part for the semi-supervised learning approach.
In the mentioned paper they mention: _. Since all of the labeled examples are interdependent, computing exact
gradients for labeled examples requires full batch gradient descent on the labeled data_ - which would somehow required seperate dataloaders/train loop for supervised/unsupervised data parts.

Is it possible to setup a mixed train loop (target/no-target in same mini-batch) and handle the missing targets somehow?

Any help would be greatly appreciated - I think this might lead to a great example for the docs

**Comment by tawe141** [comment]
Cool, I happen to be implementing something based on that exact paper :) The way I'm interpreting this is you need the entirety of the labelled data for each iteration, but you can use mini-batches of unlabeled data. So really you'd only need a dataloader for your unlabelled set.

**Comment by ackbar03** [comment]
Hi, 

I'm trying to do the same thing, the original code base they provided is too difficult to work with.  

Did any of you manage to implement the paper using gpytorch? If so, can I ask how we can minimize the variance as part of the loss function? The output of the DKL is a distribution, is its variance parameter differentiable and directly minimizable as part of the objective function?

Thanks
